clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
 
     return eraseAnonNamespace(ArgOS.str().str());
   }
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
```

Comment:
If I am understanding this correctly, you are replacing the FullyQualifiedType because it causes type mismatch, but you still need fully qualified canonical name. Did I get that right? Does it make sense to move the calculation of fully qualified canonical name outside SemaSycl?

Reply:
Yes. That's  right. Not sure about putting outside of SemaSycl.

Reply:
I don't think it makes sense to move it outside of SemaSYCL.  The way this is printed is entirely the responsiblity of the integration header logic.

Reply:
The reason I suggested it is because it looks like the required functionality here is - 'Print a canonical name with it's namespace qualification'. I didn't think that necessarily had to be restricted as a integration header functionality. Thought it might be useful in other workflows as well. That said, if it doesn't make sense to move it out, I think you can move this logic out to a separate function accepting type and calculate the required qualified name. getKernelNameTypeString() can then just call this where ever the name is required. This will also resolve the issue being discussed in L2163

Reply:
I see what you mean.  I'm actually surprised the pretty-printer doesn't do that already now that you mention it.  I doubt we should be iterating through the decl-contexts, when that likely already happens elsewhere.

Reply:
Yea me too. Makes me wonder if there is some less obvious way to get this functionality from Clang API.